[surepetcare] Sure Petcare Binding initial contribution

renescherer opened the pull request:
This is a PR for a first version of a new binding which supports cloud-connected cat flaps from Sure Petcare.
The code is quite basic and follows a similar pattern to the Weather Underground design with a regular polling processes from the Sure Petcare API.

A more detailed list of features can be found in the README.md.

I've also create a community thread for further discussions on improvements and changes:
[https://community.openhab.org/t/new-sure-petcare-binding-for-cat-pet-flap/82257](https://community.openhab.org/t/new-sure-petcare-binding-for-cat-pet-flap/82257)

This PR replaces a previous submitted one (https://github.com/openhab/openhab2-addons/pull/6112), which had to be removed due to branch issues.

renescherer commented:
I think the code has now all the key feature and should be ready to be reviewed and merged.

Hilbrand commented on bundles/org.openhab.binding.surepetcare/src/test/java/org/openhab/binding/surepetcare/internal/data/SurePetcarePetLocationTest.java:
```diff
@@ -0,0 +1,48 @@
+
+public class SurePetcarePetLocationTest {
+
+    private static final Gson gson = new GsonBuilder()
```
You should use the actual Gson used in the binding. Otherwise you might end up with a test that succeeds, but fails because the actual gson initialized in the binding is different. Possible solution put a static instance in a constants class that is used everywhere.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,489 @@
+            return pet.getPetStatus();
+        } else {
+            return null;
+        }
```
fixed by Holger

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,489 @@
+                        throw new SurePetcareApiException();
+                    }
+                }
+            } catch (Exception e) {
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/data/SurePetcarePetFeeding.java:
```diff
@@ -0,0 +1,80 @@
+    @SerializedName("device_id")
+    private Integer deviceId;
+    @SerializedName("change")
+    private List<Float> feedChange = new ArrayList<Float>();
```
why would you not use a typed ArrayList?

Hilbrand replied:
This isn't about typing or not, but to use, what is called, the diamond operator `<>`

renescherer commented on bundles/org.openhab.binding.surepetcare/src/test/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareAPIHelperTest.java:
```diff
@@ -0,0 +1,32 @@
+    public void testGetDeviceId() {
+
+        SurePetcareAPIHelper api = new SurePetcareAPIHelper();
+        api.getDeviceId();
```
fixed with Mokito mock objects

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/resources/ESH-INF/thing/things.xml:
```diff
@@ -0,0 +1,638 @@
+			<channel id="product" typeId="productType" />
+			<channel id="ledMode" typeId="ledModeType" />
+			<channel id="pairingMode" typeId="pairingModeType" />
+			<channel id="serialNumber" typeId="serialNumberType" />
```
If I use properties, can I still display the content in the sitemap?

renescherer replied:
moved several more static data points from channels to properties (e.g. hardware version, product type id etc.)

renescherer commented:
> Thanks for you contribution. I've done a partial review, but didn't want to withhold my first review comments. Some comments apply to multiple places, but I didn't logged them all. So please check if some comments might apply to other places as well.

Thanks @Hilbrand  for the detailed code review. I think I've addressed all the issues you've raised.

HerzScheisse commented:
Hey all,

(local windows VS code ide setup running OH Snapshot 2.5 1706)

I found a problem with translations here, but i don't know if this is related to this binding or a problem in general...

I have the .item & .things & binding jar files in there folders. then i clean the cache and tmp folder, start openhab via VS code IDE. OH and the binding starts correctly.

If i went to PaperUI -> Control, all the items are translated to german.

BUT there are only these items missing with item type Number:Mass

So i reloaded the .items file, now the missing items get picked up... but all the items in Paper UI -> Control are NOT translated anymore...

Anyone have a clue what is causing this problem and how to fix?

HerzScheisse commented:
hey,
still have the problem described here https://github.com/openhab/openhab2-addons/pull/6160#issuecomment-546796907

@renescherer @Hilbrand do you guys know why this happens and have maybe a fix?

renescherer commented:
Hi @Hilbrand 
I just wanted to check what else I can do to help getting this binding into the official distribution. I think all the points you raised in your code review have been closed out.
Thanks,
Rene

renescherer commented:
Apologies, I shouldn't have merged the latest 2.5.x changes into the branch. Will revert to previous commit.

Hilbrand commented:
I've changed the pr to be merged with 2.5.x so it will be available in 2.5.x. Can you please rebase on 2.5.x branch and update the pom version to `2.5.2-SNAPSHOT`

renescherer commented:
great. I'm just changing pom.xml version. When you say "rebase on 2.5.x branch", do you mean merge the 2.5.x branch into my feature/surepetcare_binding branch? If so, do I squash all the 2.5.x commits into one? Sorry, I'm still getting the hang of how to use git.

renescherer commented:
I think I've worked it out. Just do a "git rebase 2.5.x"

Hilbrand commented:
Yes that is the correct. It will just rewind your commits, apply the new commits from 2.5.x and then reapply your commits. You need to do a `push --force-with-lease` to get those changes back to GitHub.

renescherer commented:
I've rebased the branch onto 2.5.x.

tnemrap commented:
What is the status of this?

cpmeister commented:
@renescherer Are you still interested in merging this? I can start reviewing if you are ready.

HerzScheisse commented:
Hey @cpmeister 
thank you for asking... I dont know if @renescherer have time to take a look into and work on this PR, i hope he has...
But if not... what are the stepps needed to get this to an official binding... Me and a few other OH users would like to see this as oh binding...

I worked on this binding too... and have this running smoothly for quite a few months... so the binding should be almost ready, except for your review comments.

It would be nice if you can have a look into this... and if Rene dont have time... i'll could create a new PR (If thats the correct way to get this included)

Thanks
Holger

cpmeister commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,491 @@
+    private String getResultFromApi(String url) throws SurePetcareApiException {
+        boolean success = false;
+        String responseData = "";
+        while (!success) {
```
I don't like the possibility of this looping endlessly. Can you add a retry limit?

renescherer replied:
Good point. I've changed the code to retry 3 times only.

cpmeister commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,491 @@
+     * @param jsonPayload the JSON string
+     * @throws SurePetcareApiException
+     */
+    private void postDataThroughAPI(String url, HttpMethod method, String jsonPayload) throws SurePetcareApiException {
```
Can you refactor this method and `getResultFromApi` into a single method?

renescherer replied:
I've refactored the 2 methods, so that they both use the same code for the http API execution.

cpmeister commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,259 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            synchronized (petcareAPI) {
```
why are you synchronized here?

renescherer replied:
Good point. Not required anymore.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/data/SurePetcareDeviceStatus.java:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @author Rene Scherer - Initial contribution
+ */
+public class SurePetcareDeviceStatus {
```
moved all DTO classes into .dto package

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareHandlerFactory.java:
```diff
@@ -0,0 +1,103 @@
+
+    private final Logger logger = LoggerFactory.getLogger(SurePetcareHandlerFactory.class);
+
+    // private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
```
lost it :-)

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareHandlerFactory.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Reference
+    protected void setHttpClientFactory(HttpClientFactory httpClientFactory) {
+        petcareAPI.setHttpClient(httpClientFactory.getCommonHttpClient());
+    }
+
+    protected void unsetHttpClientFactory(HttpClientFactory httpClientFactory) {
+        petcareAPI.setHttpClient(null);
+    }
```
I'm not sure myself, but all other bindings with the same pattern uses "protected" scope. E.g. bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyHandlerFactory.java

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,491 @@
+     * @param curfewList the list of curfews
+     * @throws SurePetcareApiException
+     */
+    public void setCurfews(SurePetcareDevice device, SurePetcareDeviceCurfewList curfewList)
```
looks like the synchronized keyword has been missed out here. Fixed now.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,491 @@
+        int decimal = DEFAULT_DEVICE_ID;
+        try {
+            if (interfaces.hasMoreElements()) {
+                NetworkInterface netif = interfaces.nextElement();
```
I'm checking further down for a valid MAC address. Loopback interfaces don't return a MAC.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/discovery/SurePetcareDiscoveryService.java:
```diff
@@ -0,0 +1,180 @@
+    private void householdDiscovered(SurePetcareHousehold household) {
+        logger.debug("Discovered household: {}", household.getName());
+        ThingUID thingsUID = new ThingUID(THING_TYPE_HOUSEHOLD, bridgeUID, household.getId().toString());
+        Map<String, Object> properties = new HashMap<String, Object>();
+        properties.putAll(household.getThingProperties());
```
Yes. Otherwise I get the following error: The method withProperties(Map<String,Object>) in the type DiscoveryResultBuilder is not applicable for the arguments (Map<String,String>).
In theory, I could define the map as <String, Object> but then I loose type safety elsewhere.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+        this.petcareAPI = petcareAPI;
+    }
+
+    @SuppressWarnings("null")
```
The compiler doesn't like:
           if (topologyPollingJob == null || topologyPollingJob.isCancelled()) {
I've now refactored it to:
           boolean noJob = true;
            if (topologyPollingJob != null) {
                noJob = topologyPollingJob.isCancelled();
            }
            if (noJob) {

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+            Map<String, String> properties = null;
+            ThingHandler handler = th.getHandler();
+            if (handler != null) {
+                if (th.getThingTypeUID().equals(THING_TYPE_PET)) {
```
fixed.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+    protected synchronized void updateThings() {
+        logger.debug("Updating {} connected things", ((Bridge) thing).getThings().size());
+        // update existing things
+        for (Thing th : ((Bridge) thing).getThings()) {
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+            if (th.getThingTypeUID().equals(THING_TYPE_PET)) {
+                logger.debug("Updating pet status for: {}", th.getUID().getId());
+                ThingHandler handler = th.getHandler();
+                if (handler != null) {
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,235 @@
+                    if (household != null) {
+                        properties = household.getThingProperties();
+                    }
+                } else if ((th.getThingTypeUID().equals(THING_TYPE_HUB_DEVICE))
+                        || (th.getThingTypeUID().equals(THING_TYPE_FLAP_DEVICE))
+                        || (th.getThingTypeUID().equals(THING_TYPE_FEEDER_DEVICE))) {
+                    ((SurePetcareDeviceHandler) handler).updateThing();
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,262 @@
+                        updateState(PET_CHANNEL_FEEDER_DEVICE, new StringType(device.getName()));
+                        int numBowls = feeding.getFeedChange().size();
+                        for (int i = 0; (i < 2) && (i < numBowls); i++) {
+                            if (device.getControl().getBowls().getBowlId().equals(1)) {
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,262 @@
+                } else if (pet.getPetStatus().getActivity().getUserId() != null) {
+                    SurePetcareHousehold user = petcareAPI.getHousehold(pet.getHouseholdId().toString());
+                    if (user != null) {
+                        int numUsers = user.getHouseholdUsers().size();
+                        for (int i = 0; (i < numUsers); i++) {
+                            if (pet.getPetStatus().getActivity().getUserId()
+                                    .equals(user.getHouseholdUsers().get(i).getUser().getUserId())) {
+                                updateState(PET_CHANNEL_LOCATION_CHANGED_THROUGH, new StringType(
+                                        user.getHouseholdUsers().get(i).getUser().getUserName().toString()));
+                            }
+                        }
```
changed to stream processsing

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/data/SurePetcarePetActivity.java:
```diff
@@ -0,0 +1,93 @@
+    public SurePetcarePetActivity(Integer location, Date since) {
+        this.where = location;
+        this.since = since;
+    }
```
It's used in the unit test SurePetcarePetLocationTest to easily create activities.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,259 @@
+                    break;
+                case DEVICE_CHANNEL_LED_MODE:
+                    if (command instanceof StringType) {
+                        synchronized (petcareAPI) {
```
To avoid changing any value through the API while it gets refreshed by the refresh polling thread.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,259 @@
+            switch (channelUID.getId()) {
+                case DEVICE_CHANNEL_LOCKING_MODE:
+                    if (command instanceof StringType) {
+                        synchronized (petcareAPI) {
```
To avoid changing any value through the API while it gets refreshed by the refresh polling thread.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,262 @@
+                                    logger.debug("Received new location: {}", currentLocation == 1 ? 2 : 1);
+                                    // We set the location to the opposite state.
+                                    // We also set location to INSIDE (1) if currentLocation is Unknown (0)
+                                    if (commandId == 10) {
+                                        petcareAPI.setPetLocation(pet, currentLocation == 1 ? 2 : 1, new Date(
+                                                System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(commandId)));
+                                    } else if (commandId == 30) {
+                                        petcareAPI.setPetLocation(pet, currentLocation == 1 ? 2 : 1, new Date(
+                                                System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(commandId)));
+                                    } else if (commandId == 60) {
+                                        petcareAPI.setPetLocation(pet, currentLocation == 1 ? 2 : 1, new Date(
+                                                System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(commandId)));
+                                    }
```
fixed

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,262 @@
+                    if (device != null) {
+                        updateState(PET_CHANNEL_FEEDER_DEVICE, new StringType(device.getName()));
+                        int numBowls = feeding.getFeedChange().size();
+                        for (int i = 0; (i < 2) && (i < numBowls); i++) {
```
Need to check with @HerzScheisse .  I'm not quite sure what it does as I haven't got a pet feeder.

HerzScheisse replied:
I'll write you a message in the community forum with some API results.

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/data/SurePetcarePet.java:
```diff
@@ -0,0 +1,216 @@
+    private SurePetcarePhoto photo;
+
+    @SerializedName("status")
+    private SurePetcarePetStatus status = new SurePetcarePetStatus();
```
done

cpmeister commented:
Please fix these warnings:
```
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/discovery/SurePetcareDiscoveryService.java:[114,4242] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:[100,4511] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:[117,5416] Potential null pointer access: this expression has a '@Nullable' type
```

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,238 @@
+
+    private final Logger logger = LoggerFactory.getLogger(SurePetcareBridgeHandler.class);
+
+    private SurePetcareAPIHelper petcareAPI;
```
done

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareHousehold.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class SurePetcareHousehold extends SurePetcareBaseObject {
+
+    public class HouseholdUsers {
```
done

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareLoginResponse.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class SurePetcareLoginResponse {
+
+    public class Data {
```
done

renescherer commented on bundles/org.openhab.binding.surepetcare/dictionary.txt:
```diff
@@ -0,0 +1,3 @@
+petcare
+rene
```
got added to git by mistake. I've removed it now.

renescherer commented:
@Hilbrand Hi Hilbrand. Are you still interested in reviewing this PR? @cpmeister has reviewed it and is happy for it to be merged.

renescherer commented:
@Hilbrand Please could you review again or remove yourself as a reviewer, so that this binding can be merged. Many thanks, Rene

Hilbrand commented on CODEOWNERS:
```diff
@@ -183,7 +183,7 @@
 /bundles/org.openhab.binding.spotify/ @Hilbrand
 /bundles/org.openhab.binding.squeezebox/ @digitaldan @mhilbush
-/bundles/org.openhab.binding.modbus.sunspec/ @mrbig
+/bundles/org.openhab.binding.surepetcare/ @renescherer @HerzScheisse
```
You should not replace other binding entries

renescherer replied:
this has been fixed

Hilbrand commented on bom/openhab-addons/pom.xml:
```diff
@@ -911,7 +911,7 @@
     <dependency>
       <groupId>org.openhab.addons.bundles</groupId>
-      <artifactId>org.openhab.binding.modbus.sunspec</artifactId>
+      <artifactId>org.openhab.binding.surepetcare</artifactId>
```
You should not replace other binding entries

renescherer replied:
this has been fixed now.

HerzScheisse commented:
Hi :) @kaikreuzer @Hilbrand @cpmeister 

Whats the status of this binding? We know that the reviews need some time, and thats okay... Is there something that needs to be done? any chance to get this binding into the 2.5.6 release?

Thanks

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/resources/ESH-INF/thing/things.xml:
```diff
@@ -0,0 +1,619 @@
+		<description>A Sure Petcare pet feeder device</description>
+
+		<channels>
+			<channel id="id" typeId="idType" />
```
Please check the channels you expose, it seems you expose a lot, consider marking some (more) advanced:

> Especially for complex devices with a lot of channels, only a small set of channels - the most important ones - should be shown to the user to reduce complexity. Whether a channel should be declared as advanced depends on the device and can be decided by the binding developer. If a functionality is rarely used it should be better marked as advanced.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/utils/GsonColonDateTypeAdapter.java:
```diff
@@ -0,0 +1,93 @@
+    private final ColonDateFormat dateFormat;
+
+    public GsonColonDateTypeAdapter() {
+        dateFormat = new ColonDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
```
This does not look like a complex one to me, should it be some complex?

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/utils/ByteArrayFileCache.java:
```diff
@@ -0,0 +1,306 @@
+/**
+ * This is a simple file based cache implementation.
+ *
+ * @author Christoph Weitkamp - Initial contribution
```
Did @cweitkamp contribute it here?

HerzScheisse replied:
I added this for the pet picture cache... and this file was taken from the OWM binding... so i left it as is to credit the author

martinvw replied:
@cweitkamp what do you think how should we handle this?

cweitkamp replied:
I am fine with it. But we have to add myself with an "Also-By" line to the commit message.

FTR: This is the fifth binding which makes use of this util class (see https://github.com/openhab/openhab-addons/issues/4736). You can find the most recent version of it in the Chromecast binding - introduced in https://github.com/openhab/openhab-addons/pull/7367). You may want to pick that one.

I will add this binding to the list in #4736. I once tried to move it over to OHC but it did not work nicely. My current implementation uses `ConfigConstants.getUserDataFolder()` which is not available in package `org.openhab.core.cache`. I have to solve one of the todos first. There are even bigger plans for it (see e.g. eclipse-archived/smarthome#6041) to make it configurable and it could be used for voice add-ons to cache similar speech data.

HerzScheisse replied:
Thanks.
i will update this file with the newer one.

HerzScheisse replied:
Hey @cweitkamp regarding the "Also-by" that we need to the commit message... 

Can you please check on this commit if i've done it right... Thanks

https://github.com/HerzScheisse/openhab-addons/commit/a46ce12d41a5bad5c88c72e306b8dbf8d0dbbffd

cweitkamp replied:
With removal of this file you can change the sign-off again. It will not be necessary to mention me in this PR 😉 .

renescherer replied:
this uses now the common class in the core

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/utils/ByteArrayFileCache.java:
```diff
@@ -0,0 +1,306 @@
+            BigInteger bigInt = new BigInteger(1, md5Hash);
+            String fileNameHash = bigInt.toString(16);
+            // Now we need to zero pad it if you actually want the full 32 chars
+            while (fileNameHash.length() < 32) {
```
I suppose you can this without concatening a string in while loop

Hilbrand replied:
Something like this:
```
return String.format("%0"+ (32 - fileNameHash.length()) + "d%s", 0, fileNameHash));
```

HerzScheisse replied:
done

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/utils/ByteArrayFileCache.java:
```diff
@@ -0,0 +1,306 @@
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class ByteArrayFileCache {
```
What is so important and/or big that you need to cache on disk?

HerzScheisse replied:
its for the pet picture...

some testers of my original JSR223 rules script implementation have aloooot of cats (>10) so i thought this would be helpful to cache the images...

martinvw replied:
Let's keep it for now. Maybe @cweitkamp already started adding the feature to core so we can use it after migrating to OH 3

cweitkamp replied:
See https://github.com/openhab/openhab-core/pull/1723

:wink:

cweitkamp replied:
https://github.com/openhab/openhab-core/pull/1723 has been merged. You now are able to use the OHC class instead.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,251 @@
+                }
+                SurePetcarePetFeeding feeding = pet.status.feeding;
+                if (feeding != null) {
+                    SurePetcareDevice device = petcareAPI.getDevice(feeding.deviceId.toString());
```
Consider splitting this method up its very big and the nesting is also extreme.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,251 @@
+                                } catch (NumberFormatException e) {
+                                    logger.warn("Invalid location id: {}, ignoring command", commandIdStr);
+                                } catch (SurePetcareApiException e) {
+                                    logger.warn("Error from SurePetcare API. Can't update location {} for pet {}",
```
Consider logging the stacktrace, maybe on debug? IO related exceptions happen when communicating with online services and are worth a warn log, see also: 

> to log exceptions in case of temporary problems, like connection problems. In case of such exceptions this should be reflected in an updated state of the binding
[[source](https://www.openhab.org/docs/developer/guidelines.html#f-logging)]

Should you update the thing status?

martinvw replied:
Please apply this also to the methods above.

HerzScheisse replied:
hmmm... good point... i*ll think about this and do some tests

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,251 @@
+                                            new DateTimeType(pet.status.activity.getLocationChanged()));
+                                    updateState(PET_CHANNEL_LOCATION_TIMEOFFSET, UnDefType.UNDEF);
+                                } catch (NumberFormatException e) {
+                                    logger.warn("Invalid location id: {}, ignoring command", commandIdStr);
```
This should be a debug log I suppose

martinvw replied:
Please apply this also to the methods above.

HerzScheisse replied:
The location change function of this binding is the most used one... and we'd like to keep this at warn level... so the user can instantly see whats the problem without setting log level to debug...
If you still want me to change this i will do. just let me know.

martinvw replied:
Okay, let's leave this one but maybe add some more context if possible/applicable. Did you check other `warn`-logs for their importance?

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,251 @@
+                    }
+                    break;
+                case PET_CHANNEL_LOCATION_TIMEOFFSET:
+                    logger.debug("Received location time offset update command: {}", command.toFullString());
```
Consider removing the call to `toFullString` the implemenation of toString is not much different and with the current code  the command `command.toFullString()` will always be executed even when debug logging is not enabled.

martinvw replied:
Apply this also to other applicable locations

HerzScheisse replied:
done. changed in 2 locations

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcarePetHandler.java:
```diff
@@ -0,0 +1,251 @@
+
+    private static final String JPEG_CONTENT_TYPE = "image/jpeg";
+
+    private static final ByteArrayFileCache IMAGE_CACHE = new ByteArrayFileCache("org.openhab.binding.surepetcare");
```
Would it maybe make more sense to just store the actual images in the folder, that seems more optimal to me?

HerzScheisse replied:
I've used this way because we dont have to take care of that specific folder... and if a user deletes the cache... all "old images" are gone... I'd like to keep this... (update to newer version as mentioned below)

If you want me to change this... do you have a binding in mind where a can take a look how to implement this?

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,254 @@
+                        petcareAPI.setCurfews(device, existingControl.curfewList);
+                        updateThingCurfews(device);
+                    } catch (SurePetcareApiException e) {
+                        logger.warn("Error from SurePetcare API. Can't update curfews for device {}", device);
```
Can you validate whether you correctly update the thing status were needed and decreased verbosity of logging in line with https://www.openhab.org/docs/developer/guidelines.html#f-logging

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,254 @@
+                    default:
+                        break;
+                }
+                logger.debug("requiresUpdate: {}", requiresUpdate);
```
Can you either remove this or make it more meaningfully, now it looks like some remains of debugging code.

You could, for example, add inside the if a log that you are about to start setting the curfews.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,254 @@
+            SurePetcareDevice device = petcareAPI.getDevice(thing.getUID().getId());
+            if (device != null) {
+                SurePetcareDeviceControl existingControl = device.control;
+                logger.debug("Old Curfew Control: {}", existingControl);
```
Please take a look at your debug logs a lot of them seem without much context and specific to a debugging session it would be nice to make them more future proof.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,254 @@
+                channelUID.getIdWithoutGroup().length() - 1);
+        int slot = Integer.parseInt(channelUID.getAsString().substring(channelUID.getAsString().length() - 1));
+
+        synchronized (petcareAPI) {
```
Did you synchronize all of it or is it just this part?

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareDeviceHandler.java:
```diff
@@ -0,0 +1,254 @@
+                                    new QuantityType<Mass>(bowlSettings.targetId, SIUnits.GRAM));
+                        } else if (device.control.bowls.bowlId.equals(4)) {
+                            if ((i + 1) == 1) {
+                                updateState(DEVICE_CHANNEL_BOWLS_FOOD_LEFT,
```
This code seems to share logic with the `SurePetcareHandler`, could there be less duplication

HerzScheisse replied:
I'm not sure if i understand this correct...
I guess you mean the `SurePetcarePetHandler`?! If so... no these are different channels...

You have a food type and a target(weight) for each device... and in the PetHandler its the current/remaining weight...

martinvw replied:
okay, let's focus on the other things

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,239 @@
+        } else {
+            switch (channelUID.getId()) {
+                case BRIDGE_CHANNEL_REFRESH:
+                    if ("ON".equals(command.toString())) {
```
It seems you are comparing with the `OnOffType.ON.toString()`better compare with `OnOffType.ON`

HerzScheisse replied:
done

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,239 @@
+                updateState(BRIDGE_CHANNEL_ONLINE, OnOffType.ON);
+                updateThings();
+            } catch (AuthenticationException e) {
+                updateStatus(ThingStatus.OFFLINE);
```
I would expect a ThingStatus.Details and message here

HerzScheisse replied:
done

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,239 @@
+                    "@text/offline.conf-error-missing-username-or-password");
+        }
+
+        if (config.getRefreshIntervalTopology() != null) {
```
Consider splitting these parts up in different methods, it seems you already divided it into parts by putting it in separate methods you can give it a name and it will be more readable and easier to maintain for others as well.

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,239 @@
+    private final Logger logger = LoggerFactory.getLogger(SurePetcareBridgeHandler.class);
+
+    private final SurePetcareAPIHelper petcareAPI;
+    private @Nullable ScheduledFuture<?> topologyPollingJob = null;
```
This explicit ` = null` is not needed without it the same would happen

HerzScheisse replied:
done

martinvw commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBaseObjectHandler.java:
```diff
@@ -0,0 +1,83 @@
+@NonNullByDefault
+public abstract class SurePetcareBaseObjectHandler extends BaseThingHandler {
+
+    private static final int UPDATE_THING_CACHE_TIMEOUT = 3000; // 3 secs
```
You can also pass a duration to the expiring cache then you don't need the comment but can actually code it :-)

HerzScheisse replied:
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/README.md:
```diff
@@ -0,0 +1,384 @@
+    Text item=UR_1e_Weight icon="text"
+    Text item=UR_1e_DateOfBirth
+    Text item=UR_1e_FeedDevice
+    /*Text item=UR_1e_FeedChange icon="text"*/
```
There are 2 tyes of bowls (1 big or 2 small bowls) this item is used for the big bowl... 
If the user is using the 2 small bowls, the following 2 items are used (UR_1e_FeedChangeLeft, UR_1e_FeedChangeRight)

martinvw replied:
Maybe you could add a comment/text to explain which option you should use in what case. For me, it was not immediately clear.

HerzScheisse replied:
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/resources/ESH-INF/thing/things.xml:
```diff
@@ -0,0 +1,619 @@
+
+	<channel-type id="feederLastChangeLeftType">
+		<item-type>Number:Mass</item-type>
+		<label>Pet Feeding Last Change Left</label>
```
I fear i dont understand you comment...

This value is for the pet feeder device... it has a scale included... so if the pet eat and left the device... the change in gram is reported.

martinvw replied:
> data 

Uhm, sorry I meant 'date'. I thought the label sounded like a date but the unit suggested otherwise. If you have the idea that you could improve the title that would of course be great.

HerzScheisse commented:
Thank you @martinvw for the code review... I added some comments for the parts i added to the binding... 
Hopefully @renescherer find some time to work on your review comments. :)

martinvw commented:
Hello @renescherer how are you doing? I hope you are fine. Please let me know if you have any questions, I have a lot of time to respond this week, after that there are two weeks in which I have less options to reply.

Thanks a lot for your time and efforts.

marioaltvater commented:
I was following this PR since several months and am still waiting to finally see this addon in OH.
Not being an engineer myself, I am definitely puzzled why the process of adding addons seems so overly complicated and in this particular case it was dragged out this much.
@renescherer @HerzScheisse if you find the energy to bring it over the finish line, I am sure there are many more users looking forward to your work. 🙏

HerzScheisse commented:
> I was following this PR since several months and am still waiting to finally see this addon in OH.
> Not being an engineer myself, I am definitely puzzled why the process of adding addons seems so overly complicated and in this particular case it was dragged out this much.
> @renescherer @HerzScheisse if you find the energy to bring it over the finish line, I am sure there are many more users looking forward to your work. 🙏

Hey @marioaltvater and others...

sorry for the long delay... but i have had no time to work on this...

Now i found some spare hours... and have finally my branch up to date with the latest openhab/2.5.x  branch and a working IDE again...

i will try to get as many review comments solved as possible... keep in mind that i'm not a Java Dev... and my need some help/guidance here and there... and provide a 2.5.9/10 binding...

The next big step is then to migrate this to OH3... I already tried but cant find a way to push my 2.5.x branch to the openhab/main branch... and i must say i'm not as good on github as others here... maybe someone can give me a hint or How-To to solve this... I already read alot information online and on the OH community... but failed to move this...

Thanks.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/README.md:
```diff
@@ -0,0 +1,384 @@
+
+The binding code is based on a lot of work done by other developers:
+
+- Holger Eisold (https://github.com/HerzScheisse) - Python use in OpenHAB and various PRs (https://github.com/HerzScheisse/SurePetcare-openHAB-JSR223-Rules)
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/config/items/bridge.items:
```diff
@@ -0,0 +1,3 @@
+Group   dgPet                                               
```
you are right... deleted those files.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDeviceCurfewList.java:
```diff
@@ -0,0 +1,89 @@
+     */
+    @Override
+    public SurePetcareDeviceCurfew get(int index) {
+        if (index >= size()) {
```
the flap can have up to 4 curfew times... the API then returns the times for each.

martinvw replied:
But why does it need to grow when you try to fetch something. If you don't want to handle null it might make more sense to return an empty object instead of modifying the list. It might be unexpected that get actually modifies the list.

But given the time you have available, it might make more sense to leave this as is, given that it works now, and focus on the error handling / thing-status updating. Because this issue is about polishing they other or more important.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareLoginResponse.java:
```diff
@@ -0,0 +1,48 @@
+    public Data data;
+
+    @Override
+    public String toString() {
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcarePetActivity.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class SurePetcarePetActivity {
+
+    @SerializedName("tag_id")
```
can you explain what you mean and where do i find/edit this setting? Thanks

martinvw replied:
https://stackoverflow.com/a/11810183/3656774

I expect some other openHAB bindings will also use it.

HerzScheisse replied:
thanks...
okay i got/understood this... and we still use this FieldNamingPolicy in the binding... I removed all the unneeded `@SerializedName` where possible...

Some are atill present... since we changed the fields in the binding....
Is there a negative impact to use this `@SerializedName` annotation? or can we leave them for now?

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcarePetActivity.java:
```diff
@@ -0,0 +1,53 @@
+    public Long deviceId;
+    @SerializedName("user_id")
+    public Long userId;
+    @SerializedName("where")
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareUser.java:
```diff
@@ -0,0 +1,63 @@
+    // }
+    // }
+
+    public String emailAddress;
```
done.
Also fixed this in others files.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/handler/SurePetcareBridgeHandler.java:
```diff
@@ -0,0 +1,239 @@
+    public void initialize() {
+        logger.debug("Initializing Sure Petcare bridge handler.");
+        SurePetcareBridgeConfiguration config = getConfigAs(SurePetcareBridgeConfiguration.class);
+        updateState(BRIDGE_CHANNEL_ONLINE, OnOffType.OFF);
```
the API reports this status of the device itself...
So how do we preceed here? rename this channel to something else?

martinvw replied:
You could use it to update the thing-status, so that would mean that you remove this line.

And when an update is received from the API mentioning that a device is OFFLINE, you could use `updateStatus` with sub status communication-error and a description informing the user that the device is reported OFFLINE by the API.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcarePet.java:
```diff
@@ -0,0 +1,104 @@
+    public String name = "";
+
+    @SerializedName("gender")
+    public Integer genderId;
```
okay

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/resources/ESH-INF/thing/things.xml:
```diff
@@ -0,0 +1,619 @@
+			<channel id="product" typeId="productType" />
+			<channel id="ledMode" typeId="ledModeType" />
+			<channel id="pairingMode" typeId="pairingModeType" />
+			<channel id="online" typeId="onlineType" />
```
See above:
https://github.com/openhab/openhab-addons/pull/6160#discussion_r504322098

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDeviceCurfew.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class SurePetcareDeviceCurfew {
+
+    public Boolean enabled;
```
does this still apply here? if so what need to be set to boolean?

martinvw replied:
a `Boolean` in Java is an object, so it can be `true`, `false`, and `null` but using it `null` does not seem to make much sense in this context. So by using a boolean you make it explicit that will be either `true` or `false`

HerzScheisse replied:
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,477 @@
+    private String password = "";
+    private boolean online = false;
+
+    private @NonNullByDefault({}) HttpClient httpClient;
```
can you please tell me which line? its hard to tell to which line the review comment is related to... Thanks

martinvw replied:
That is about the line `private @NonNullByDefault({}) HttpClient httpClient;`it is just assigned once so to make that more clear apparent you can make it final.

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,477 @@
+                logger.debug("current hostname: {}, device id: {}", hostname, decimal);
+            }
+        } catch (SocketException e) {
+            logger.debug("Socket Exception: {}", e.getMessage());
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/discovery/SurePetcareDiscoveryService.java:
```diff
@@ -0,0 +1,181 @@
+                break;
+            case UNKNOWN:
+            default:
+                break;
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareBaseObject.java:
```diff
@@ -0,0 +1,55 @@
+@NonNullByDefault
+public class SurePetcareBaseObject {
+
+    public Long id = Long.valueOf(0);
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareBaseObject.java:
```diff
@@ -0,0 +1,55 @@
+    public Date createdAt = new Date();
+    public Date updatedAt = new Date();
+
+    public SurePetcareBaseObject() {
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareLoginCredentials.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public SurePetcareLoginCredentials(String emailAddress, String password, String deviceId) {
+        super();
```
done

HerzScheisse commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java:
```diff
@@ -0,0 +1,477 @@
+                }
+            } catch (AuthenticationException | InterruptedException | ExecutionException | TimeoutException
+                    | ProtocolException e) {
+                logger.debug("Exception caught during API execution: {}", e.getMessage());
```
done

renescherer commented:
@Hilbrand , @martinvw 

Unfortunately I wasn't able to work on this for the last few months, but I've got now time to close out the last remaining issues.
 
Since this PR has been open for quite a while, would it make sense to close this and start a new PR and start straight away on the latest 3.x branch? As @HerzScheisse mentioned as well, I'm not sure how otherwise to move this PR onto a new branch and would need a bit of help.

Many thanks!

HerzScheisse commented:
Hey Rene @renescherer 

Nice to hear you are back... hope everything is fine...

Me didnt have had time to finish the review but i did a couple of things... the "latest" version is here in my tree... if you want to have a look

https://github.com/HerzScheisse/openhab-addons/tree/surepetcare

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareTopology.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    public @Nullable SurePetcareHousehold getHouseholdById(String id) {
+        for (SurePetcareHousehold household : households) {
```
done

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareBridgeConfiguration.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class SurePetcareBridgeConfiguration {
+
+    private String username;
```
done

renescherer commented on bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareHousehold.java:
```diff
@@ -0,0 +1,49 @@
+        return "SurePetcareHousehold [id=" + id + ", name=" + name + "]";
+    }
+
+    public class HouseholdUsers {
```
I've made them static. What is the benefit of having them as independent classes? The HouseholdUsers class is a wrapper around the actual user infomation. 
        "users":[
          {
            "id":52815,
            "owner":true,
            "write":true,
            "version":"MA==",
            "created_at":"2019-09-02T08:20:45+00:00",
            "updated_at":"2019-09-02T08:20:50+00:00",
            "user":{
              "id":52815,
              "name":"Rene Scherer"
            }
          },
...
        ]
      }
    ]

renescherer commented:
I've migrated the code base to OH3 of the main branch and therefore close this PR.

A replacement PR (https://github.com/openhab/openhab-addons/pull/9713) has been raised.
